dll/win32/iernonce/registry.cpp

```diff
@@ -0,0 +1,200 @@
+    _In_ const CStringW &lpSubKeyName) :
+    m_SectionName(lpSubKeyName)
+{
+    m_bSuccess = FALSE;
+    CRegKeyEx hKey;
+    DWORD dwValueNum;
+    DWORD dwMaxValueNameLen;
+    LSTATUS Error;
+    CStringW ValueName;
+
+    if (hKey.Open(hParentKey, lpSubKeyName, KEY_READ) != ERROR_SUCCESS)
+        return;
+
+    Error = RegQueryInfoKeyW(hKey, NULL, 0, NULL, NULL, NULL, NULL,
+                             &dwValueNum, &dwMaxValueNameLen,
+                             NULL, NULL, NULL);
+    if (Error != ERROR_SUCCESS)
+        return;
+
+    for (DWORD i = 0; i < dwValueNum; i++)
+    {
+        LPWSTR szValueName;
+        DWORD dwcchName = dwMaxValueNameLen + 1;
+
+        szValueName = ValueName.GetBuffer(dwMaxValueNameLen + 1);
+        Error = hKey.EnumValueName(i, szValueName, &dwcchName);
+        ValueName.ReleaseBuffer();
+
+        if (Error != ERROR_SUCCESS)
+        {
+            // TODO: error handling
+            return;
+        }
+
+        if (!HandleValue(hKey, ValueName, dwcchName + 1))
+            return;
+    }
+
+    // sort entries by name in string order.
+    qsort(m_EntryList.GetData(), m_EntryList.GetSize(), sizeof(RunOnceExEntry),
+          RunOnceExEntryCmp);
+
+    hKey.Close();
+    m_bSuccess = TRUE;
+    return;
+}
+
+int RunOnceExSectionCmp(_In_ const void *a, _In_ const void *b)
+{
+    return lstrcmpW(((RunOnceExSection *)a)->m_SectionName,
+                    ((RunOnceExSection *)b)->m_SectionName);
+}
+
+RunOnceExInstance::RunOnceExInstance(_In_ HKEY BaseKey)
+{
+    m_bSuccess = FALSE;
+    CRegKeyEx hKey;
+    DWORD dwSubKeyNum;
+    DWORD dwMaxSubKeyNameLen;
+    LSTATUS Error;
+    CStringW SubKeyName;
+
+    if (hKey.Open(BaseKey,
+                  L"Software\\Microsoft\\Windows\\CurrentVersion\\RunOnceEx\\",
+                  KEY_READ) != ERROR_SUCCESS)
+    {
+        return;
+    }
+
+    Error = RegQueryInfoKeyW(hKey, NULL, 0, NULL,
+                             &dwSubKeyNum, &dwMaxSubKeyNameLen,
+                             NULL, NULL, NULL, NULL, NULL, NULL);
+    if (Error != ERROR_SUCCESS)
+        return;
+
+    for (DWORD i = 0; i < dwSubKeyNum; i++)
+    {
+        LPWSTR szSubKeyName;
+        DWORD dwcchName = dwMaxSubKeyNameLen + 1;
+
+        szSubKeyName = SubKeyName.GetBuffer(dwMaxSubKeyNameLen + 1);
+        Error = hKey.EnumKey(i, szSubKeyName, &dwcchName);
+        SubKeyName.ReleaseBuffer();
+
+        if (Error != ERROR_SUCCESS)
+        {
+            // TODO: error handling
+            return;
+        }
+
+        if (!HandleSubKey(hKey, SubKeyName, dwcchName + 1))
+            return;
+    }
+
+    // sort sections by name in string order.
+    qsort(m_SectionList.GetData(), m_SectionList.GetSize(), sizeof(RunOnceExSection),
+          RunOnceExSectionCmp);
+
+    hKey.Close();
+    m_bSuccess = TRUE;
+    return;
+}
+
+BOOL RunOnceExInstance::HandleSubKey(
+    _In_ CRegKeyEx &hKey,
+    _In_ const CStringW& SubKeyName,
+    _In_ DWORD cchszName)
```

Comment:
`cchszName` is unused, what was this for?

Reply:
oh, I forgot it's already zero-terminated, the length is not necessary!